src/gui/src/layoutTabs.cpp

```diff
@@ -0,0 +1,327 @@
+    current_viewer_->startRulerBuild();
+  }
+}
+
+void LayoutTabs::cancelRulerBuild()
+{
+  if (current_viewer_) {
+    current_viewer_->cancelRulerBuild();
+  }
+}
+
+void LayoutTabs::selection(const Selected& selection)
+{
+  if (current_viewer_) {
+    current_viewer_->selection(selection);
+  }
+}
+
+void LayoutTabs::selectionFocus(const Selected& focus)
+{
+  if (current_viewer_) {
+    current_viewer_->selectionFocus(focus);
+  }
+}
+
+void LayoutTabs::updateModuleVisibility(odb::dbModule* module, bool visible)
+{
+  modules_[module].visible = visible;
+  fullRepaint();
+}
+
+void LayoutTabs::updateModuleColor(odb::dbModule* module,
+                                   const QColor& color,
+                                   bool user_selected)
+{
+  modules_[module].color = color;
+  if (user_selected) {
+    modules_[module].user_color = color;
+  }
+  fullRepaint();
+}
+
+void LayoutTabs::populateModuleColors(odb::dbBlock* block)
+{
+  if (block == nullptr) {
+    return;
+  }
+
+  ColorGenerator generator;
+
+  for (auto* module : block->getModules()) {
+    auto color = generator.getQColor();
+    modules_[module] = {color, color, color, true};
+  }
+}
+
+void LayoutTabs::addFocusNet(odb::dbNet* net)
+{
+  const auto& [itr, inserted] = focus_nets_.insert(net);
+  if (inserted) {
+    emit focusNetsChanged();
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::addRouteGuides(odb::dbNet* net)
+{
+  const auto& [itr, inserted] = route_guides_.insert(net);
+  if (inserted) {
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::addNetTracks(odb::dbNet* net)
+{
+  const auto& [itr, inserted] = net_tracks_.insert(net);
+  if (inserted) {
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::removeFocusNet(odb::dbNet* net)
+{
+  if (focus_nets_.erase(net) > 0) {
+    emit focusNetsChanged();
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::removeRouteGuides(odb::dbNet* net)
+{
+  if (route_guides_.erase(net) > 0) {
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::removeNetTracks(odb::dbNet* net)
+{
+  if (net_tracks_.erase(net) > 0) {
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::clearFocusNets()
+{
+  if (!focus_nets_.empty()) {
+    focus_nets_.clear();
+    emit focusNetsChanged();
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::clearRouteGuides()
+{
+  if (!route_guides_.empty()) {
+    route_guides_.clear();
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::clearNetTracks()
+{
+  if (!net_tracks_.empty()) {
+    net_tracks_.clear();
+    fullRepaint();
+  }
+}
+
+void LayoutTabs::exit()
+{
+  for (auto viewer : viewers_) {
+    viewer->exit();
+  }
+}
+
+void LayoutTabs::commandAboutToExecute()
+{
+  if (current_viewer_) {
+    current_viewer_->commandAboutToExecute();
+  }
+}
+
+void LayoutTabs::commandFinishedExecuting()
+{
+  if (current_viewer_) {
+    current_viewer_->commandFinishedExecuting();
+  }
+}
+
+void LayoutTabs::restoreTclCommands(std::vector<std::string>& cmds)
+{
+  if (current_viewer_) {
+    current_viewer_->restoreTclCommands(cmds);
```

Comment:
this probably needs to be fixed at some point.
This code implies that only the state of the current window is saved, but you would want to be able to restore the view of all the windows correctly.
I think there is probably a better way than the tclcommands but that is for a different time

Reply:
Agreed but I'll look at it later